### PR TITLE
pkg/uroot: rewrite bb to be more general

### LINKED ISSUE #423: More robust bb source rewriting
Currently, for each command to be included in the final binary, bb does the following (please correct me if I missed important steps):

1. Rewrites main to Main.
2. Rewrites init to Init.
3. Creates a command flag.FlagSet `cmdflag`.
4. Rewrites `flag.Foo()` calls to `cmdflag.Foo()`.

This has a number of limitations, ordered from minor to major:

1. It only supports a single init function.

This is trivially fixable.

2. It doesn't rewrite global var assignment.

These may have side effects. This is fairly easily fixed by moving the assignments into Init.

3. It screws up local variables named `flag`.

The AST doesn't know if `flag` is a package or variable identifier.

4. It doesn't rewrite assignments to flag.Usage, breaking custom usage output.

I have a commit to fix this, but it doesn't quite work correctly because a bunch of packages set os.Arg[0] to change the usage output, but flag.FlagSet.Usage does not print os.Arg[0] like flag.Usage does.

5. It doesn't fix up any other global side effects.

This is the fundamental problem.

Imports are not rewritten, so if a command package imports another package which has side effects, it will affect every other command.

That means any flags registered in other packages will end up on `flag.CommandLine` instead of `cmdflags`.

Or that package may have init functions with other arbitrary side effects.

I think this can be solved robustly by doing away with flag-specific rewriting, and instead rewriting all source, including imported packages, to encapsulate all side effects.

For example, a package like this:

```go
package main

import (
    "fmt"

    "import1"
    "import2"
)

var foo = flag.Bool("foo", false, "bar")

func init() {
    fmt.Printf("initing")
}

func main() {
    flag.Parse()
    // Do stuff.
}
```

Becomes this:

```go
package cmd

import (
    "fmt"

    "import1"
    "import2"
)

var foo *bool

func initVars() {
    foo = flag.Bool("foo", false, "bar")
}

func init1() {
    fmt.Printf("initing")
}

var initDone bool

func Init() {
    if initDone {
        return
    }

    import1.Init()
    import2.Init()

    initVars()
    init1()

    initDone = true
}

func Main() {
    flag.Parse()
    // Do stuff.
}
```

Notice that use of the flag package no longer needs to be rewritten, because no flags will be registered (in any package) until Init is called.

The generated init is basically a recreation of the normal Go init functions, recursively calling init on all dependencies.

This approach has a few challenges:

1. We probably don't want to rebuild the standard library, so we can't rewrite it. However I think that is mostly fine. The only package I can think of with side effects on import is "testing" (adds flags).

2. Moving global variable assignment into a function requires type information. The AST alone will no longer suffice. This is needed to fix (3) above anyways.

3. The real main package in this binary needs to be written very carefully to avoid its own side effects, or using these rewritten packages without calling Init.

This is definitely trickier than the existing mechanism, but beyond not rewriting the standard library, I can't think of any cases where it would break. Please let me know what I forgot.

### PR DESCRIPTION
**Pull request by hugelgupf**
Fixes #423.

Still a work in progress, but seems to work with all of our existing ./cmds/* tools. (It compiles, I haven't tried it yet. But I'm taking that as a good sign...)

The downside is that bb mode now takes longer to create an initramfs than source mode. But it's still manageable.

cc @prattmic @rminnich @rjoleary

### CODE REVIEW

**Comment by hugelgupf** [comment]
There's a lot more to be done here... cleanup, tests, and apparently reproducibility is failing on bb-amd64.

But please try it out and report back whether stuff is failing (esp NiChrome -- I imagine there are a couple Go AST edge cases I don't have yet).

File: pkg/uroot/bb.go [hunk 1]
```diff
@@ -250,6 +254,93 @@ func (p *Package) rewriteFile(opts BuildOpts, f *ast.File) bool {
...
+			}
+			for _, spec := range d.Specs {
+				s := spec.(*ast.ValueSpec)
+				if s.Values == nil {
+					continue
+				}
+
+				// Add an assign statement for these values to init.
+				//
+				// TODO: Double-check the order in which
```
Comment (prattmic) [inline pkg/uroot/bb.go]: FYI, variables are initialized only after their dependencies: https://golang.org/ref/spec#Package_initialization

go/types.Info.InitOrder contains this ordering: https://godoc.org/go/types#Info

Comment (prattmic) [reply pkg/uroot/bb.go]: To answer the TODO, init functions always come after declarations: https://github.com/golang/go/blob/master/src/cmd/compile/internal/gc/init.go#L71-L72

Comment (hugelgupf) [reply pkg/uroot/bb.go]: :heart_eyes: thanks!

**Comment by hugelgupf** [comment]
I'm gonna be honest: if I had known rewriting bb this way was this simple, and less than 100 more lines of code, I wouldn't have resisted it so long.

**Comment by rminnich** [comment]
The acid test, I think will be NiChrome, so let me know when you're ready.

File: pkg/uroot/bb.go [hunk 2]
```diff
@@ -25,96 +28,79 @@ import (
...
-		//log.Printf("run the Init function for %v: run %v", os.Args[0], f)
 		f()
 	}
 
+	// Yeah, it's weird that this stuff happens in bbsh's init.
+	// We only do it if we actually *are* init. But merging this into bbsh
+	// should happen some other way. This should just be in bbsh.Main?
+	//
+	// Maybe init should just be a normal bb command, just like the others,
+	// except with a special symlink to /init?
```
Comment (rminnich) [inline pkg/uroot/bb.go]: but you want init and rush as one command. That's the issue. Two commands and you blow the rom space budget. The space problem is the source of many sins.

File: pkg/uroot/bb.go [hunk 3]
```diff
@@ -25,96 +28,79 @@ import (
...
+	// Set up root file system. Mount stuff.
 	util.Rootfs()
 
-        // spawn the first shell. We had been running the shell as pid 1
-        // but that makes control tty stuff messy. We think.
-        cloneFlags := uintptr(0)
+	// Figure out which init to run.
+	//
+	// Spawn the first shell. We had been running the shell as pid 1 but
+	// that makes control tty stuff messy. We think.
```
Comment (rminnich) [inline pkg/uroot/bb.go]: ok so something got lost here. The inito needs its own PID space. So the cloneflags I think ought to be set to syscall.clone_NEWPID the first time through. That means line 104 (left) is wrong but the reset of this is mostly right. We need to do some testing before we rip the cloneFlags bits out.

File: pkg/uroot/bb.go (hunk 2 above)
Comment (hugelgupf) [reply pkg/uroot/bb.go]: You do want it as one command. But why not just compile init into rush just like all the other bb commands? Everything already gets compiled into rush, no need to make init special except for the symlink.

File: pkg/uroot/bb.go (hunk 3 above)
Comment (hugelgupf) [reply pkg/uroot/bb.go]: I think right now that's out of scope for this PR -- the cloneFlags are already out, and if we want them back in, we should do some testing independent of this PR and reenable them.

**Comment by hugelgupf** [comment]
PTAL

**Comment by hugelgupf** [comment]
ping, anyone feel like reviewing?

File: pkg/uroot/bb.go [hunk 4]
```diff
@@ -223,109 +209,133 @@ func BBBuild(opts BuildOpts) (ArchiveFiles, error) {
...
 		CmdName: p.name,
-		Init:    p.init,
 	}
 }
 
-func (p *Package) rewriteFile(opts BuildOpts, f *ast.File) bool {
-	// Inspect the AST and change all instances of main()
-	var pos token.Position
+// TODO:
+// - write an init name generator, in case InitN is already taken.
```
Comment (rjoleary) [inline pkg/uroot/bb.go]: I like how Grumpy prefixes generated identifiers with a Greek symbol to indicate they've been generated and not a character a human would type. For example, ßInit. Something like that could be used here.

File: pkg/uroot/bb.go (hunk 3 above)
Comment (rjoleary) [reply pkg/uroot/bb.go]: This is the out-of-date version of the code here: https://github.com/u-root/u-root/blob/master/cmds/init/init.go#L159

File: pkg/uroot/bb.go (hunk 2 above)
Comment (hugelgupf) [reply pkg/uroot/bb.go]: Agreed. I'm actually thinking that we may want two different inits. One for source mode, one for bb mode.

```
cmds/bb/init
cmds/source/init
```

Comment (hugelgupf) [reply pkg/uroot/bb.go]: On the other hand, if we want to allow multi-mode initramfs, maybe one init should be suitable.

Comment (rjoleary) [reply pkg/uroot/bb.go]: Another possibility would be a single init with a custom buildtag gating feature we don't want in bb. For example, "bgbuild.go" which we definitely don't want in bb mode would have the line `// +build !bb`. I believe Go lets you use your own buildtags with the `ctxt.BuildTags` slice or via the `-tags` option in "go build".

File: pkg/uroot/bb.go [hunk 5]
```diff
@@ -383,18 +393,59 @@ func getPackage(opts BuildOpts, importPath string) (*Package, error) {
...
+	}
+
+	var files []*ast.File
+	for _, f := range pp.ast.Files {
+		files = append(files, f)
+	}
+	// Type-check the package before we continue. We need types to rewrite
+	// some statements.
+	conf := types.Config{
+		Importer: importer.For("source", nil),
```
[the remark this answers is not in the document]
Comment (hugelgupf) [reply pkg/uroot/bb.go]: You have a good point, but there's an easier way to shave ~40 seconds off the build. As this is, it takes ~65 seconds to build all u-root packages with bb. With the next patch set, it'll be ~25 seconds. Thanks!

File: pkg/uroot/bb.go (hunk 4 above)
Comment (hugelgupf) [reply pkg/uroot/bb.go]: It's not a general solution (people can name Go functions anything they want), so I'd rather find some other way. But maybe the default can be something like this.

File: pkg/uroot/bb.go (hunk 3 above)
Comment (hugelgupf) [reply pkg/uroot/bb.go]: Just another reason to just have one init, or to not have init baked into bb at least. I still thing that's something for a different PR so we can have a better discussion about this.

File: pkg/uroot/bb.go (hunk 2 above)
Comment (hugelgupf) [reply pkg/uroot/bb.go]: That's a nice idea. Like I said, another PR.

**Comment by hugelgupf** [comment]
ping. need to fix reproducibility, but pls take another look so i can get this in soon. @rminnich or @ganshun or @rjoleary

**Comment by hugelgupf** [comment]
I fixed the reproducibility issue and some golint stuff. Going to push this in.
